lldb-eval/parser.cc

```diff
@@ -1025,19 +1073,45 @@ void Parser::ParsePtrOperator(TypeDeclaration* type_decl) {
   }
 }
 
-bool Parser::ResolveTypeFromTypeDecl(const TypeDeclaration& type_decl) {
-  // TODO(werat): This is an over-broad simplification, because TypeDeclaration
-  // doesn't have any type validation at the moment. Types like "char char" are
-  // considered valid and built-in, yet they don't resolve to anything. For the
-  // sake of parsing it's OK, it will be handled at the evaluation step.
-  if (type_decl.is_builtin_) {
-    return true;
+lldb::SBType Parser::ResolveTypeFromTypeDecl(const TypeDeclaration& type_decl) {
+  if (!type_decl.IsValid()) {
+    return lldb::SBType();
   }
 
   // Resolve the type in the current expression context.
-  lldb::SBType type = ctx_->ResolveTypeByName(type_decl.GetBaseName().c_str());
+  return ctx_->ResolveTypeByName(type_decl.GetBaseName().c_str());
+}
+
+lldb::SBType Parser::ResolveTypeDeclarators(lldb::SBType type,
+                                            const TypeDeclaration& type_decl) {
+  // Resolve pointers/references.
+  for (clang::tok::TokenKind tk : type_decl.ptr_operators_) {
+    if (tk == clang::tok::star) {
+      // Pointers to reference types are forbidden.
+      if (type.IsReferenceType()) {
+        std::string msg = llvm::formatv(
+            "'type name' declared as a pointer to a reference of type '{0}'",
+            type.GetName());
+        BailOut(ErrorCode::kInvalidOperandType, msg, token_.getLocation());
+        return lldb::SBType();
+      }
+      // Get pointer type for the base type: e.g. int* -> int**.
+      type = type.GetPointerType();
+
+    } else if (tk == clang::tok::amp) {
+      // References to references are forbidden.
+      if (type.IsReferenceType()) {
+        BailOut(ErrorCode::kInvalidOperandType,
+                "type name declared as a reference to a reference",
+                token_.getLocation());
```

Comment:
It looks like the error is a bit different from the above, is it intentional?

If not, then you can perhaps hoist the `if (type.IsReferenceType()) {` the check to the beginning of the loop.

Reply:
The error messages are the same as in `clang`.  Also these cases are different: one is pointer-to-reference, second is reference-to-reference.